proposals/record-structs.md

```diff
@@ -0,0 +1,301 @@
+
+If `Equals(R other)` is user-defined (not synthesized) but `GetHashCode` is not, a warning is produced.
+
+```C#
+public bool Equals(R other);
+```
+
+The synthesized `Equals(R)` returns `true` if and only if for each instance field `fieldN` in the record struct
+the value of `System.Collections.Generic.EqualityComparer<TN>.Default.Equals(fieldN, other.fieldN)` where `TN` is the field type is `true`.
+
+The record struct includes synthesized `==` and `!=` operators equivalent to operators declared as follows:
+```C#
+public static bool operator==(R r1, R r2)
+    => r1.Equals(r2);
+public static bool operator!=(R r1, R r2)
+    => !(r1 == r2);
+```
+The `Equals` method called by the `==` operator is the `Equals(R other)` method specified above. The `!=` operator delegates to the `==` operator. It is an error if the operators are declared explicitly.
+
+The record struct includes a synthesized override equivalent to a method declared as follows:
+```C#
+public override bool Equals(object? obj);
+```
+It is an error if the override is declared explicitly. 
+The synthesized override returns `other is R && Equals((R)other)` where `R` is the record struct.
+
+The record struct includes a synthesized override equivalent to a method declared as follows:
+```C#
+public override int GetHashCode();
+```
+The method can be declared explicitly.
+
+A warning is reported if one of `Equals(R)` and `GetHashCode()` is explicitly declared but the other method is not explicit.
+
+The synthesized override of `GetHashCode()` returns an `int` result of a deterministic function combining the values of `System.Collections.Generic.EqualityComparer<TN>.Default.GetHashCode(fieldN)` for each instance field `fieldN` with `TN` being the type of `fieldN`.
+
+For example, consider the following record struct:
+```C#
+record struct R1(T1 P1, T2 P2);
+```
+
+For this record struct, the synthesized equality members would be something like:
+```C#
+struct R1 : IEquatable<R1>
+{
+    public T1 P1 { get; init; }
+    public T1 P1 { get; init; }
+    public override bool Equals(object? obj) => obj is R1 && Equals((R1)obj);
+    public virtual bool Equals(R1 other)
+    {
+        return
+            EqualityComparer<T1>.Default.Equals(P1, other.P1) &&
+            EqualityComparer<T2>.Default.Equals(P2, other.P2);
+    }
+    public static bool operator==(R1 r1, R1 r2)
+        => r1.Equals(r2);
+    public static bool operator!=(R1 r1, R1 r2)
+        => !(r1 == r2);    
+    public override int GetHashCode()
+    {
+        return Combine(
+            EqualityComparer<T1>.Default.GetHashCode(P1),
+            EqualityComparer<T2>.Default.GetHashCode(P2));
+    }
+}
+```
+
+### Printing members: PrintMembers and ToString methods
+
+The record struct includes a synthesized method equivalent to a method declared as follows:
+```C#
+private bool PrintMembers(System.Text.StringBuilder builder);
```

Comment:
`PrintMembers` seems like overkill for a record struct, as the point was to support inheritance for record classes. I guess this would make it resilient if we ever wanted to add struct inheritance, but we also removed EqualityContract from equality so that's broken anyway. We should consider simplifying this since we don't need the abstraction.

Reply:
Yes. I did consider it and also called it out as an open question.
The reason I kept `PrintMembers` is for more symmetry between `class` and `struct` records. You can convert from one to another with less change, and the mental model of the APIs you can customize is more consistent.

---
In reply to: [553001779](https://github.com/dotnet/csharplang/pull/4307#discussion_r553001779) [](ancestors = 553001779)

Reply:
I'd have it be an open question then. I can see the argument, but given that the only use for these members is to support inheritance I think it's not worth the overhead.

Reply:
Yes. I already have this open question to cover this: "- confirm that we want to keep PrintMembers design (separate method returning `bool`)"

---
In reply to: [553689611](https://github.com/dotnet/csharplang/pull/4307#discussion_r553689611) [](ancestors = 553689611)